src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1179,22 +1180,41 @@
 (pls/defn-validated deactivate-node!
   "Deactivate the given host, recording the current time. If the node is
   currently inactive, no change is made."
-  [certname :- String]
-  (sql/do-prepared "UPDATE certnames SET deactivated = ?
+  [certname :- String & [timestamp :- pls/Timestamp]]
+  (let [timestamp (to-timestamp (or timestamp (now)))]
+   (sql/do-prepared "UPDATE certnames SET deactivated = ?
                     WHERE certname=? AND deactivated IS NULL"
-                   [(to-timestamp (now)) certname]))
+                    [timestamp certname])))
+
+
+(defn- timestamp-of-newest-record [entity certname time-field]
+  (let [query {:select [time-field]
+               :from [entity]
+               :where [:= :certname certname]
+               :order-by [[time-field :desc]]
+               :limit 1}]
+    (sql/with-query-results result-set (hcore/format query)
+      (time-field (first result-set)))))
+
+(pls/defn-validated have-record-produced-after?
+  [entity :- s/Keyword
+   certname :- String
+   time :- pls/Timestamp]
+  (let [time (to-timestamp time)
+        time-field (if (= entity :reports) :start_time :producer_timestamp)]
```

Comment:
A few things:
- receive_time is definitely out, as it's from from the local clock.

- No matter what we do, it needs to converge in the same way no matter what
order the messages are received and processed in. start_time and end_time
are the choices.

- If an admin is logged into the master and tries to deactivate a node
while, at the same time, a puppet run is happening on that node, what
should happen? We can either prefer the deactivate or prefer the puppet
run. I think it makes sense to prefer the deactivate since it was
explicitly indicated by the user (i.e. use start_time), but I can't say I'm
really invested in that viewpoint.


On Wed, Apr 22, 2015 at 3:51 PM, Wyatt Alt <notifications@github.com> wrote:

> In src/puppetlabs/puppetdb/scf/storage.clj
> <https://github.com/puppetlabs/puppetdb/pull/1361#discussion_r28922577>:
>
> > +
> > +(defn- timestamp-of-newest-record [entity certname time-field]
> > +  (let [query {:select [time-field]
> > +               :from [entity]
> > +               :where [:= :certname certname]
> > +               :order-by [[time-field :desc]]
> > +               :limit 1}]
> > +    (sql/with-query-results result-set (hcore/format query)
> > +      (time-field (first result-set)))))
> > +
> > +(pls/defn-validated have-record-produced-after?
> > +  [entity :- s/Keyword
> > +   certname :- String
> > +   time :- pls/Timestamp]
> > +  (let [time (to-timestamp time)
> > +        time-field (if (= entity :reports) :start_time :producer_timestamp)]
>
> there may be some subtlety here around using start_time vs end_time. If
> the puppet run takes a very long time for some reason, the two could be
> quite different. I'm not sure what the right choice is exactly, but I think
> there's a possibility here that a puppet run starts at t=0, deactivate node
> issued at t=1, puppet run completes at t=2, report processed at t=3, and
> node deactivate is processed at t=4. I think my expectation under this
> situation would be for the deactivate to fail (because a report has come
> in), but as written here I believe it would would succeed. receive_time
> would be yet another candidate timestamp. I'm not sure that it's wrong as
> written, I think I just need a bit more clarity on the intent.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetdb/pull/1361/files#r28922577>.
>